lib/SIL/IR/SILInstructions.cpp

```diff
@@ -727,18 +727,6 @@ DifferentiableFunctionExtractInst::DifferentiableFunctionExtractInst(
                                : getExtracteeType(function, extractee, module),
                            function.getOwnershipKind()),
       Extractee(extractee), HasExplicitExtracteeType(extracteeType.hasValue()) {
-#ifndef NDEBUG
-  if (extracteeType.hasValue()) {
-    // Note: explicit extractee type is used to avoid inconsistent typing in:
-    // - Canonical SIL, due to generic specialization.
-    // - Lowered SIL, due to LoadableByAddress.
-    // See `TypeSubstCloner::visitDifferentiableFunctionExtractInst` for an
-    // explanation of how explicit extractee type is used.
-    assert((module.getStage() == SILStage::Canonical ||
-            module.getStage() == SILStage::Lowered) &&
-           "Explicit type is valid only in canonical or lowered SIL");
-  }
-#endif
```

Comment:
Relaxing this assertion isn't so nice, but I would appreciate being able to land it to unblock a use case unless a different solution is apparent.

A more robust longer-term solution may be to change SIL `@differentiable` function types to explicitly store component original/JVP/VJP function types, instead of inferring JVP/VJP types from the original type. @rxwei had this idea long ago.

That would seem to solve this issue, along with enabling the following:
- Robustly re-enabling `differentiable_function` instruction verification (TF-1197).
- Enabling straightforward cloning/transformation of `differentiable_function_extract`
  instructions without [worrying about breaking invariants](https://github.com/apple/swift/blob/949b0c0cc3632b0c278a33a4022dd280cf31215a/include/swift/SIL/TypeSubstCloner.h#L325-L348).